File: python/game_of_life/game.py
class Game():
    def __init__(self, size=0, seed=None):
        self.size = len(seed) if seed is not None else size
        self.current_generation = self.__create_dead_generation()

        if seed is not None:
            self.__seed_board(seed)

    def next_generation(self):
        next_generation = self.__create_dead_generation()
        for row_index in range(self.size):
            for column_index in range(self.size):
                next_generation[row_index][column_index] = self.__get_next_state(row_index, column_index)
        return next_generation

    def __count_living_neighbors(self, row_index, column_index):
        living_count = 0
        living_count += 1 if self.__is_upper_left_alive(row_index, column_index) else 0
        living_count += 1 if self.__is_upper_alive(row_index, column_index) else 0
        living_count += 1 if self.__is_upper_right_alive(row_index, column_index) else 0

        living_count += 1 if self.__is_left_alive(row_index, column_index) else 0
        living_count += 1 if self.__is_right_alive(row_index, column_index) else 0

        living_count += 1 if self.__is_lower_left_alive(row_index, column_index) else 0
        living_count += 1 if self.__is_lower_alive(row_index, column_index) else 0
        living_count += 1 if self.__is_lower_right_alive(row_index, column_index) else 0
        return living_count

    def __is_upper_left_alive(self, row_index, column_index):
        return self.__is_alive(row_index - 1, column_index - 1)

    def __is_upper_alive(self, row_index, column_index):
        return self.__is_alive(row_index - 1, column_index)

    def __is_upper_right_alive(self, row_index, column_index):
        return self.__is_alive(row_index - 1, column_index + 1)

    def __is_left_alive(self, row_index, column_index):
        return self.__is_alive(row_index, column_index - 1)

    def __is_right_alive(self, row_index, column_index):
        return self.__is_alive(row_index, column_index + 1)

    def __is_lower_left_alive(self, row_index, column_index):
        return self.__is_alive(row_index + 1, column_index - 1)

    def __is_lower_alive(self, row_index, column_index):
        return self.__is_alive(row_index + 1, column_index)

    def __is_lower_right_alive(self, row_index, column_index):
        return self.__is_alive(row_index + 1, column_index + 1)

    def __is_alive(self, row_index, column_index):
        return -1 < row_index < self.size \
               and -1 < column_index < self.size \
               and self.current_generation[row_index][column_index]

    def __seed_board(self, seed):
        for row_index in range(len(seed)):
            for column_index in range(len(seed[row_index])):
                self.current_generation[row_index][column_index] = seed[row_index][column_index]

    def __create_dead_generation(self):
        return [[False for i in range(self.size)] for i in range(self.size)]

    def __get_next_state(self, row_index, column_index):
        living_neighbors = self.__count_living_neighbors(row_index, column_index)

        if self.__is_alive(row_index, column_index):
            return 2 <= living_neighbors < 4

        return living_neighbors == 3

File: ruby/lib/game.rb
class Game
  attr_reader :current_generation


  def initialize(size: nil, seed: nil)
    @current_generation = create_dead_generation(size) unless size.nil?
    @current_generation = seed unless seed.nil?
  end

  def size
    @current_generation.size
  end

  def next_generation
    next_gen = create_dead_generation(size)
    (0..size - 1).each { |row_index|
      (0..size - 1).each { |column_index|
        next_gen[row_index][column_index] = should_live_or_revive(row_index, column_index)
      }
    }
    next_gen
  end

  private
  def should_live_or_revive(row_index, column_index)
    return should_revive(row_index, column_index) if is_dead(row_index, column_index)
    should_live(row_index, column_index)
  end

  def should_live(row_index, column_index)
    living_neighbors = count_living_neighbors(row_index, column_index)
    return false if living_neighbors < 2
    return false if living_neighbors > 3
    true
  end

  def should_revive(row_index, column_index)
    living_neighbors = count_living_neighbors(row_index, column_index)
    living_neighbors == 3
  end

  def count_living_neighbors(row_index, column_index)
    living_neighbors = 0
    living_neighbors += 1 if is_upper_left_alive(row_index, column_index)
    living_neighbors += 1 if is_upper_alive(row_index, column_index)
    living_neighbors += 1 if is_upper_right_alive(row_index, column_index)

    living_neighbors += 1 if is_left_alive(row_index, column_index)
    living_neighbors += 1 if is_right_alive(row_index, column_index)

    living_neighbors += 1 if is_lower_left_alive(row_index, column_index)
    living_neighbors += 1 if is_lower_alive(row_index, column_index)
    living_neighbors += 1 if is_lower_right_alive(row_index, column_index)

    living_neighbors
  end

  def is_lower_right_alive(row_index, column_index)
    is_living(row_index + 1, column_index + 1)
  end

  def is_lower_alive(row_index, column_index)
    is_living(row_index + 1, column_index)
  end

  def is_lower_left_alive(row_index, column_index)
    is_living(row_index + 1, column_index - 1)
  end

  def is_right_alive(row_index, column_index)
    is_living(row_index, column_index + 1)
  end

  def is_left_alive(row_index, column_index)
    is_living(row_index, column_index - 1)
  end

  def is_upper_right_alive(row_index, column_index)
    is_living(row_index - 1, column_index + 1)
  end

  def is_upper_alive(row_index, column_index)
    is_living(row_index - 1, column_index)
  end

  def is_upper_left_alive(row_index, column_index)
    is_living(row_index - 1, column_index - 1)
  end

  def is_living(row_index, column_index)
    is_in_generation(row_index, column_index) && @current_generation[row_index][column_index]
  end

  def is_dead(row_index, column_index)
    not @current_generation[row_index][column_index]
  end

  def is_in_generation(row_index, column_index)
    row_index > -1 && row_index < size && column_index > -1 && column_index < size
  end

  def create_dead_generation(size)
    Array.new(size, Array.new(size, false))
  end
end